### ISSUE TITLE
By default, Node `SupabaseClient` leaks itself due to `SupabaseAuthClient` `setInterval`

### ISSUE THREAD

**Issue by mqp** [supabase/supabase-js#926]
# Bug report

- [x] I confirm this is a bug with Supabase, not with my own application. 
- [x] I confirm I have searched the [Docs](https://docs.supabase.com), GitHub [Discussions](https://github.com/supabase/supabase/discussions), and [Discord](https://discord.supabase.com). 

## Describe the bug

This bit us the other day:

1. We construct a temporary `SupabaseClient` somewhere, like in an Express endpoint.
2. By default, the `SupabaseClient` constructs a `SupabaseAuthClient` with `autoRefreshToken = true`.
3. The `SupabaseAuthClient` calls `_handleVisibilityChange`, which calls `startAutoRefresh`, which creates a `setInterval` closing over the `SupabaseAuthClient` object.
4. We finish our request and the `SupabaseClient` goes out of scope.
5. We didn't know we had to call `client.auth.stopAutoRefresh()`, so the interval is still running and the whole thing is leaked forever doing useless work.
6. Oops, our Express app does this on every request, bad news for our server.

I'm not sure what the desired fix here would be, but it seems like a big foot gun for this to happen by default. Obviously if something like [this proposal](https://github.com/tc39/proposal-explicit-resource-management) was in Javascript then you could expose a `dispose` method and people would figure out that they really needed to call `dispose`, but as of 2023 that's probably not really going to stop people from getting burned by this.

(Unsure whether this should be in the `supabase-js` repo or the `gotrue-js` repo since it's unclear which layer should be changed.)

## Screenshots

![image](https://github.com/supabase/supabase-js/assets/428637/2ea301e9-f4cd-4fbf-b198-4573698a7385)

**Comment by j4w8n** [comment]
I'm guessing you're not using an auth helper package to create the client.

If using the native `createClient` function, you'll want to set these auth options for that client.

```js
autoRefreshToken: false,
detectSessionInUrl: false,
persistSession: false
```

This is assuming that you're not keeping track of a user session with this client.

**Comment by mqp** [comment]
Yep, setting `autoRefreshToken` to `false` works for us. My only complaint is that the default failed silently in a nasty way and there wasn't really any documentation calling out what I needed to do in a server context for it to be OK, so it seems like there could be room for improvement.

**Comment by j4w8n** [comment]
Yeah I think there used to be a small blurb about this in the docs, but I can't find it now.

They could set it up to be automagic based on a browser context - this is what auth helpers/ssr does - but that would be a breaking change for the native client function.

**Comment by kangmingtay** [comment]
@mqp these are good points - perhaps we could use something like [`FinalizationRegistry`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/FinalizationRegistry#description) to register a cleanup method when the class is instantiated on the server-side.

although it seems like they have a huge warning too :/ (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/FinalizationRegistry#avoid_where_possible)

**Comment by mandarini** [comment]
I will be closing this issue since it has no activity for more than 2 years.

If you find that the issue persists on the latest version of our libs, feel free to open a new one, ideally with a small reproduction repository.

Thank you for your patience and for contributing to Supabase!
